src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersReservedSlotsBehavior.cs

```diff
@@ -0,0 +1,150 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Configuration.Settings;
+using Stratis.Bitcoin.Connection;
+using Stratis.Bitcoin.Consensus;
+using Stratis.Bitcoin.Interfaces;
+using Stratis.Bitcoin.P2P;
+using Stratis.Bitcoin.P2P.Peer;
+using Stratis.Bitcoin.P2P.Protocol;
+using Stratis.Bitcoin.P2P.Protocol.Behaviors;
+using Stratis.Bitcoin.P2P.Protocol.Payloads;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus.Behaviors
+{
+   public class ProvenHeadersReservedSlotsBehavior : NetworkPeerBehavior
+   {
+      private readonly IConnectionManager connectionManager;
+      private readonly ILoggerFactory loggerFactory;
+      private readonly ConsensusSettings consensusSettings;
+      private readonly ICheckpoints checkpoints;
+      private readonly IChainState chainState;
+
+      /// <summary>
+      /// The proven header peers reserved slots threshold (%).
+      /// Represents the percentage of maximum connectable peers that we reserve for peers that are able to serve proven headers.
+      /// </summary>
+      private const decimal ProvenHeaderPeersReservedSlotsThreshold = 0.4M;
+
+      /// <summary>Instance logger.</summary>
+      private readonly ILogger logger;
+
+      public ProvenHeadersReservedSlotsBehavior(
+         IConnectionManager connectionManager,
+         ILoggerFactory loggerFactory,
+         ConsensusSettings consensusSettings,
+         ICheckpoints checkpoints,
+         IChainState chainState)
+      {
+         this.connectionManager = connectionManager;
+         this.loggerFactory = loggerFactory;
+         this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
+         this.consensusSettings = consensusSettings;
+         this.checkpoints = checkpoints;
+         this.chainState = chainState;
+      }
+
+      /// <summary>
+      /// Processes and incoming message from the peer.
+      /// </summary>
+      /// <param name="peer">Peer from which the message was received.</param>
+      /// <param name="message">Received message to process.</param>
+      protected async Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+      {
+         switch (message.Message.Payload)
```

Comment:
no need to do switch case here, if will be sufficient

Reply:
In a previous behavior i used a straight if on a single condition but Dan asked me to use switch.
https://github.com/stratisproject/StratisBitcoinFullNode/pull/2403#discussion_r223388625

Please guys, agree on this and update the code style guidline if this is a subject of PR reviews.

Reply:
We use switch in all other places lets stay consistent